framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,531 @@
+                printAnnotations(it.tlAnnotationsHere);
+                printWriter.append(' ');
+            }
+        }
+    }
+
+    /**
+     * Prints the stub representation of an AField, which represents a variable declaration. In
+     * practice, this should either be a field or a method parameters, since there should be no
+     * local variable declarations in a stub.
+     *
+     * @param aField the field to print
+     * @param fieldName the name to use for the declaration in the stub file. This doesn't matter
+     *     for parameters, but must be correct for fields.
+     * @param className the name of the enclosing class. This is only used for printing the type of
+     *     an explicit receiver parameter (i.e. a parameter named "this").
+     */
+    private void printAField(AField aField, String fieldName, String className) {
+        String basetype;
+        if ("this".equals(fieldName)) {
+            basetype = className;
+        } else if (basetypes.containsKey(aField.type.description.toString())) {
+            basetype = basetypes.get(aField.type.description.toString()).toString();
+        } else {
+            basetype = "TODOTYPE";
+        }
+
+        if (basetype.contains("[")) {
+            String component = basetype.substring(0, basetype.lastIndexOf('['));
+            printArrayComponentTypeAnnotation(aField.type);
+            printWriter.print(component);
+            printWriter.print(" ");
+            basetype = "[]";
+        }
+        printAnnotations(aField.type.tlAnnotationsHere);
+        printWriter.print(" ");
+        printWriter.print(basetype);
+        printWriter.print(" ");
+        printWriter.print(fieldName);
+    }
+
+    /**
+     * Writes out an import statement for each annotation used in an {@link AScene}.
+     *
+     * <p>{@code DefCollector} is a facility in the Annotation File Utilities for determining which
+     * annotations are used in a given AScene. Here, we use that construct to write out the proper
+     * import statements into a stub file.
+     */
+    private class ImportDefCollector extends DefCollector {
+
+        /**
+         * Constructs a new ImportDefCollector, which will run on the given AScene when its {@code
+         * visit} method is called.
+         *
+         * @param scene the scene whose imported annotations should be printed
+         * @throws DefException if the DefCollector does not succeed
+         */
+        ImportDefCollector(AScene scene) throws DefException {
+            super(scene);
+        }
+
+        /**
+         * Write an import statement for a given AnnotationDef. This is only called once per
+         * annotation used in the scene.
+         *
+         * @param d the annotation definition to print an import for
+         */
+        @Override
+        protected void visitAnnotationDef(AnnotationDef d) {
+            if (!d.name.contains("+")) {
+                printWriter.println("import " + d.name + ";");
```

Comment:
I'm concerned that conflicting imports are possible.  For example, code might already use JSR 305 nullness annotations, then infer-and-annotate might insert Checker Framework nullness annotations.
Can you add a test for such a case?

Reply:
> I'm concerned that conflicting imports are possible. For example, code might already use JSR 305 nullness annotations, then infer-and-annotate might insert Checker Framework nullness annotations.

Good idea. I added a test. It passes, but it did require me to get rid of a lot of wildcard imports in the test cases, so a win all around.